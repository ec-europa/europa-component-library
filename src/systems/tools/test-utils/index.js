const merge = require('deepmerge');
const twing = require('@ecl-twig/ec-storybook/.storybook/environment');

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) => {
    try {
      const html = twing.render(file, options);
      const test = document.createElement('test');
      test.innerHTML = html.trim();
      resolve(test);
    } catch (error) {
      reject(error);
    }
  });

module.exports = {
  merge,
  renderTwigFileAsNode,
};
