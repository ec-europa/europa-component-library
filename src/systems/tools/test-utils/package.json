{
  "name": "@ecl-twig/test-utils",
  "version": "2.31.1",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
