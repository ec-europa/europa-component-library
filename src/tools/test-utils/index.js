const merge = require('deepmerge');
const twing = require('@ecl/twig-ec-storybook/.storybook/environment');

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) => {
    try {
      const html = twing.render(file, options);
      const div = document.createElement('div');
      div.innerHTML = html.trim();
      resolve(div.firstChild);
    } catch (error) {
      reject(error);
    }
  });

module.exports = {
  merge,
  renderTwigFileAsNode,
};
