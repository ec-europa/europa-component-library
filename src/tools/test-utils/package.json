{
  "name": "@ecl/test-utils",
  "version": "3.4.2",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
