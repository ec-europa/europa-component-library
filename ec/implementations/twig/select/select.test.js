import { merge, renderTwigFileAsNode } from '@ecl-twig/test-utils';

import dataSingle from './demo/data--single';
import dataMultiple from './demo/data--multiple';

describe('EC - Select', () => {
  const template = '@ecl-twig/ec-component-select/ecl-select.html.twig';
  const render = (params) => renderTwigFileAsNode(template, params);

  describe('Default', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(
        render(
          merge(dataSingle, {
            required: false,
          })
        )
      ).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(dataSingle, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const optionsWithExtraAttributes = merge(dataSingle, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(
        render(optionsWithExtraAttributes)
      ).resolves.toMatchSnapshot();
    });

    test('with missing input data and debug enabled returns the right warning message', () => {
      expect.assertions(1);

      const dataCompliance = { ...dataSingle, _compliance_: true };
      dataCompliance.options[0].value = '';

      return expect(render(dataCompliance)).resolves.toMatchSnapshot();
    });
  });

  describe('Required', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(dataSingle)).resolves.toMatchSnapshot();
    });
  });

  describe('Disabled', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(
        render(
          merge(dataSingle, {
            disabled: true,
            required: false,
          })
        )
      ).resolves.toMatchSnapshot();
    });
  });

  describe('With error', () => {
    const options = merge(dataSingle, {
      invalid: true,
      required: false,
    });

    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });

  describe('Multiple', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(dataMultiple)).resolves.toMatchSnapshot();
    });

    test('with missing input data and debug enabled returns the right warning message', () => {
      expect.assertions(1);

      const dataCompliance = { ...dataMultiple, _compliance_: true };
      dataCompliance.multiple_all_text = '';
      dataCompliance.multiple_search_text = '';
      dataCompliance.multiple_placeholder = '';

      return expect(render(dataCompliance)).resolves.toMatchSnapshot();
    });
  });
});
