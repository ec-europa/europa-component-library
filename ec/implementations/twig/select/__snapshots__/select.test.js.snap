// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EC - Select Default renders correctly 1`] = `
<test>
  <div
    class="ecl-form-group"
  >
    <label
      class="ecl-form-label"
      for="select-default"
    >
      Select a country
      <span
        class="ecl-form-label__optional"
      >
          (optional)
      </span>
    </label>
    <div
      class="ecl-help-block"
    >
      This is the input's helper text.
    </div>
    <div
      class="ecl-select__container ecl-select__container--m"
    >
      <select
        class="ecl-select"
        id="select-default"
      >
        <option
          value="1"
        >
          Belgium
        </option>
        <option
          value="2"
        >
          France
        </option>
        <option
          value="3"
        >
          Luxembourg
        </option>
        <option
          value="4"
        >
          Germany
        </option>
        <option
          value="5"
        >
          Bulgaria
        </option>
        <option
          value="6"
        >
          Italy
        </option>
        <option
          value="7"
        >
          Romania
        </option>
        <option
          value="8"
        >
          Greece
        </option>
        <option
          value="9"
        >
          Hungary
        </option>
        <option
          value="10"
        >
          Portugal
        </option>
      </select>
      <div
        class="ecl-select__icon"
      >
        <svg
          aria-hidden="true"
          class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
          focusable="false"
        >
          <use
            xlink:href="/icons.svg#ui--corner-arrow"
          />
        </svg>
      </div>
    </div>
  </div>
</test>
`;

exports[`EC - Select Default renders correctly with extra attributes 1`] = `
<test>
  <div
    class="ecl-form-group"
  >
    <label
      class="ecl-form-label"
      for="select-default"
    >
      Select a country
      <span
        class="ecl-form-label__required"
      >
        *
      </span>
    </label>
    <div
      class="ecl-help-block"
    >
      This is the input's helper text.
    </div>
    <div
      class="ecl-select__container ecl-select__container--m"
    >
      <select
        class="ecl-select"
        data-test="data-test-value"
        data-test-1="data-test-value-1"
        id="select-default"
        required=""
      >
        <option
          value="1"
        >
          Belgium
        </option>
        <option
          value="2"
        >
          France
        </option>
        <option
          value="3"
        >
          Luxembourg
        </option>
        <option
          value="4"
        >
          Germany
        </option>
        <option
          value="5"
        >
          Bulgaria
        </option>
        <option
          value="6"
        >
          Italy
        </option>
        <option
          value="7"
        >
          Romania
        </option>
        <option
          value="8"
        >
          Greece
        </option>
        <option
          value="9"
        >
          Hungary
        </option>
        <option
          value="10"
        >
          Portugal
        </option>
      </select>
      <div
        class="ecl-select__icon"
      >
        <svg
          aria-hidden="true"
          class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
          focusable="false"
        >
          <use
            xlink:href="/icons.svg#ui--corner-arrow"
          />
        </svg>
      </div>
    </div>
  </div>
</test>
`;

exports[`EC - Select Default renders correctly with extra class names 1`] = `
<test>
  <div
    class="ecl-form-group"
  >
    <label
      class="ecl-form-label"
      for="select-default"
    >
      Select a country
      <span
        class="ecl-form-label__required"
      >
        *
      </span>
    </label>
    <div
      class="ecl-help-block"
    >
      This is the input's helper text.
    </div>
    <div
      class="ecl-select__container ecl-select__container--m"
    >
      <select
        class="ecl-select custom-class custom-class--test"
        id="select-default"
        required=""
      >
        <option
          value="1"
        >
          Belgium
        </option>
        <option
          value="2"
        >
          France
        </option>
        <option
          value="3"
        >
          Luxembourg
        </option>
        <option
          value="4"
        >
          Germany
        </option>
        <option
          value="5"
        >
          Bulgaria
        </option>
        <option
          value="6"
        >
          Italy
        </option>
        <option
          value="7"
        >
          Romania
        </option>
        <option
          value="8"
        >
          Greece
        </option>
        <option
          value="9"
        >
          Hungary
        </option>
        <option
          value="10"
        >
          Portugal
        </option>
      </select>
      <div
        class="ecl-select__icon"
      >
        <svg
          aria-hidden="true"
          class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
          focusable="false"
        >
          <use
            xlink:href="/icons.svg#ui--corner-arrow"
          />
        </svg>
      </div>
    </div>
  </div>
</test>
`;

exports[`EC - Select Default with missing input data and debug enabled returns the right warning message 1`] = `
<test>
  <div
    class="ecl-form-group"
  >
    <label
      class="ecl-form-label"
      for="select-default"
    >
      Select a country
      <span
        class="ecl-form-label__required"
      >
        *
      </span>
    </label>
    <div
      class="ecl-help-block"
    >
      This is the input's helper text.
    </div>
    <div
      class="ecl-select__container ecl-select__container--m"
    >
      <select
        class="ecl-select"
        id="select-default"
        required=""
      >
        <option
          value=""
        >
          Belgium
        </option>
        <option
          value="2"
        >
          France
        </option>
        <option
          value="3"
        >
          Luxembourg
        </option>
        <option
          value="4"
        >
          Germany
        </option>
        <option
          value="5"
        >
          Bulgaria
        </option>
        <option
          value="6"
        >
          Italy
        </option>
        <option
          value="7"
        >
          Romania
        </option>
        <option
          value="8"
        >
          Greece
        </option>
        <option
          value="9"
        >
          Hungary
        </option>
        <option
          value="10"
        >
          Portugal
        </option>
      </select>
      <div
        class="ecl-select__icon"
      >
        <svg
          aria-hidden="true"
          class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
          focusable="false"
        >
          <use
            xlink:href="/icons.svg#ui--corner-arrow"
          />
        </svg>
      </div>
    </div>
  </div>
  <span
    class="ecl-u-type-color-green ecl-compliance-inner ecl-compliance-success ecl-u-ma-2xs"
    title="[form-group--select] - ecl compliant"
  >
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--xs"
      focusable="false"
    >
      <use
        xlink:href="/icons.svg#notifications--success"
      />
    </svg>
  </span>
  <!-- ecl2-compliance: select -->
  <div
    class="ecl2-compliance ecl-u-border-width-1 ecl-u-border-color-red ecl-u-border-style-solid ecl-u-ma-xs ecl-u-pr-s ecl-u-bg-white"
  >
    <p
      class="ecl-u-type-color-grey ecl-u-ml-m ecl-u-type-xs"
    >
      [select] -  Each option needs to contain a {{ label }} and a {{ value }} properties set and filled in.
    </p>
  </div>
</test>
`;

exports[`EC - Select Disabled renders correctly 1`] = `
<test>
  <div
    class="ecl-form-group"
  >
    <label
      class="ecl-form-label ecl-form-label--disabled"
      for="select-default"
    >
      Select a country
      <span
        class="ecl-form-label__optional"
      >
          (optional)
      </span>
    </label>
    <div
      class="ecl-help-block ecl-help-block--disabled"
    >
      This is the input's helper text.
    </div>
    <div
      class="ecl-select__container ecl-select__container--disabled ecl-select__container--m"
    >
      <select
        class="ecl-select"
        disabled=""
        id="select-default"
      >
        <option
          value=""
        >
          Belgium
        </option>
        <option
          value="2"
        >
          France
        </option>
        <option
          value="3"
        >
          Luxembourg
        </option>
        <option
          value="4"
        >
          Germany
        </option>
        <option
          value="5"
        >
          Bulgaria
        </option>
        <option
          value="6"
        >
          Italy
        </option>
        <option
          value="7"
        >
          Romania
        </option>
        <option
          value="8"
        >
          Greece
        </option>
        <option
          value="9"
        >
          Hungary
        </option>
        <option
          value="10"
        >
          Portugal
        </option>
      </select>
      <div
        class="ecl-select__icon"
      >
        <svg
          aria-hidden="true"
          class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
          focusable="false"
        >
          <use
            xlink:href="/icons.svg#ui--corner-arrow"
          />
        </svg>
      </div>
    </div>
  </div>
</test>
`;

exports[`EC - Select Multiple renders correctly 1`] = `
<test>
  <div
    class="ecl-form-group"
  >
    <label
      class="ecl-form-label"
      for="select-multiple"
    >
      Select a country
      <span
        class="ecl-form-label__required"
      >
        *
      </span>
    </label>
    <div
      class="ecl-help-block"
    >
      This is the input's helper text.
    </div>
    <div
      class="ecl-select__container ecl-select__container--m"
    >
      <select
        class="ecl-select"
        data-ecl-auto-init="Select"
        data-ecl-select-all="Select all"
        data-ecl-select-default="Placeholder text"
        data-ecl-select-multiple=""
        data-ecl-select-search="Search"
        id="select-multiple"
        multiple=""
        required=""
      >
        <option
          value="1"
        >
          Belgium
        </option>
        <option
          value="2"
        >
          France
        </option>
        <option
          value="3"
        >
          Luxembourg
        </option>
        <option
          value="4"
        >
          Germany
        </option>
        <option
          value="5"
        >
          Bulgaria
        </option>
        <option
          value="6"
        >
          Italy
        </option>
        <option
          value="7"
        >
          Romania
        </option>
        <option
          value="8"
        >
          Greece
        </option>
        <option
          value="9"
        >
          Hungary
        </option>
        <option
          value="10"
        >
          Portugal
        </option>
      </select>
      <div
        class="ecl-select__icon"
      >
        <svg
          aria-hidden="true"
          class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
          focusable="false"
        >
          <use
            xlink:href="/icons.svg#ui--corner-arrow"
          />
        </svg>
      </div>
    </div>
  </div>
</test>
`;

exports[`EC - Select Multiple with missing input data and debug enabled returns the right warning message 1`] = `
<test>
  <div
    class="ecl-form-group"
  >
    <label
      class="ecl-form-label"
      for="select-multiple"
    >
      Select a country
      <span
        class="ecl-form-label__required"
      >
        *
      </span>
    </label>
    <div
      class="ecl-help-block"
    >
      This is the input's helper text.
    </div>
    <div
      class="ecl-select__container ecl-select__container--m"
    >
      <select
        class="ecl-select"
        data-ecl-auto-init="Select"
        data-ecl-select-all=""
        data-ecl-select-default=""
        data-ecl-select-multiple=""
        data-ecl-select-search=""
        id="select-multiple"
        multiple=""
        required=""
      >
        <option
          value="1"
        >
          Belgium
        </option>
        <option
          value="2"
        >
          France
        </option>
        <option
          value="3"
        >
          Luxembourg
        </option>
        <option
          value="4"
        >
          Germany
        </option>
        <option
          value="5"
        >
          Bulgaria
        </option>
        <option
          value="6"
        >
          Italy
        </option>
        <option
          value="7"
        >
          Romania
        </option>
        <option
          value="8"
        >
          Greece
        </option>
        <option
          value="9"
        >
          Hungary
        </option>
        <option
          value="10"
        >
          Portugal
        </option>
      </select>
      <div
        class="ecl-select__icon"
      >
        <svg
          aria-hidden="true"
          class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
          focusable="false"
        >
          <use
            xlink:href="/icons.svg#ui--corner-arrow"
          />
        </svg>
      </div>
    </div>
  </div>
  <span
    class="ecl-u-type-color-green ecl-compliance-inner ecl-compliance-success ecl-u-ma-2xs"
    title="[form-group--select] - ecl compliant"
  >
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--xs"
      focusable="false"
    >
      <use
        xlink:href="/icons.svg#notifications--success"
      />
    </svg>
  </span>
  <!-- ecl2-compliance: select -->
  <div
    class="ecl2-compliance ecl-u-border-width-1 ecl-u-border-color-red ecl-u-border-style-solid ecl-u-ma-xs ecl-u-pr-s ecl-u-bg-white"
  >
    <p
      class="ecl-u-type-color-grey ecl-u-ml-m ecl-u-type-xs"
    >
      [select] -  When the select is multiple the {{ multiple_placeholder }} property needs to be set and filled. When the select is multiple the {{ multiple_search_text }} property needs to be set and filled. When the select is multiple the {{ multiple_all_text }} property needs to be set and filled.
    </p>
  </div>
</test>
`;

exports[`EC - Select Required renders correctly 1`] = `
<test>
  <div
    class="ecl-form-group"
  >
    <label
      class="ecl-form-label"
      for="select-default"
    >
      Select a country
      <span
        class="ecl-form-label__required"
      >
        *
      </span>
    </label>
    <div
      class="ecl-help-block"
    >
      This is the input's helper text.
    </div>
    <div
      class="ecl-select__container ecl-select__container--m"
    >
      <select
        class="ecl-select"
        id="select-default"
        required=""
      >
        <option
          value=""
        >
          Belgium
        </option>
        <option
          value="2"
        >
          France
        </option>
        <option
          value="3"
        >
          Luxembourg
        </option>
        <option
          value="4"
        >
          Germany
        </option>
        <option
          value="5"
        >
          Bulgaria
        </option>
        <option
          value="6"
        >
          Italy
        </option>
        <option
          value="7"
        >
          Romania
        </option>
        <option
          value="8"
        >
          Greece
        </option>
        <option
          value="9"
        >
          Hungary
        </option>
        <option
          value="10"
        >
          Portugal
        </option>
      </select>
      <div
        class="ecl-select__icon"
      >
        <svg
          aria-hidden="true"
          class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
          focusable="false"
        >
          <use
            xlink:href="/icons.svg#ui--corner-arrow"
          />
        </svg>
      </div>
    </div>
  </div>
</test>
`;

exports[`EC - Select With error renders correctly 1`] = `
<test>
  <div
    class="ecl-form-group"
  >
    <label
      class="ecl-form-label ecl-form-label--invalid"
      for="select-default"
    >
      Select a country
      <span
        class="ecl-form-label__optional"
      >
          (optional)
      </span>
    </label>
    <div
      class="ecl-help-block"
    >
      This is the input's helper text.
    </div>
    <div
      class="ecl-feedback-message"
    >
      This is the error message
    </div>
    <div
      class="ecl-select__container ecl-select__container--m"
    >
      <select
        class="ecl-select ecl-select--invalid"
        id="select-default"
      >
        <option
          value="1"
        >
          Belgium
        </option>
        <option
          value="2"
        >
          France
        </option>
        <option
          value="3"
        >
          Luxembourg
        </option>
        <option
          value="4"
        >
          Germany
        </option>
        <option
          value="5"
        >
          Bulgaria
        </option>
        <option
          value="6"
        >
          Italy
        </option>
        <option
          value="7"
        >
          Romania
        </option>
        <option
          value="8"
        >
          Greece
        </option>
        <option
          value="9"
        >
          Hungary
        </option>
        <option
          value="10"
        >
          Portugal
        </option>
      </select>
      <div
        class="ecl-select__icon"
      >
        <svg
          aria-hidden="true"
          class="ecl-icon ecl-icon--s ecl-icon--rotate-180 ecl-select__icon-shape"
          focusable="false"
        >
          <use
            xlink:href="/icons.svg#ui--corner-arrow"
          />
        </svg>
      </div>
    </div>
  </div>
</test>
`;
